sqlite.go

```diff
@@ -0,0 +1,542 @@
+//
+//
+// Binding Time
+//
+// While SQLite3 has no strict time datatype, it does have a series of built-in
+// functions that operate on timestamps that expect columns to be in one of many
+// formats: https://sqlite.org/lang_datefunc.html
+//
+// When encoding a time.Time into one of SQLite's preferred formats, we use the
+// shortest timestamp that the time.Time can be represented using, one of:
+//
+//	2. YYYY-MM-DD HH:MM
+//	3. YYYY-MM-DD HH:MM:SS
+//	4. YYYY-MM-DD HH:MM:SS.SSS
+//
+// If the time.Time is not UTC (strongly consider storing times in UTC!),
+// we follow SQLite's norm of appending "[+-]HH:MM" to the above formats.
+//
+// It is common for efficiency to store a time as seconds-since-epoch in an
+// INTEGER column. If you want to do that, pass the result of time.Time.Unix
+// to the driver.
+//
+//
+// Reading Time
+//
+// In general, time is hard to extract from SQLite as a time.Time.
+// If a column is defined as DATE or DATETIME, then the text can be parsed
+// as TimeFormat and returned as a time.Time.
+package sqlite
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"encoding"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/tailscale/sqlite/cgosqlite"
+	"github.com/tailscale/sqlite/sqliteh"
+)
+
+// TimeFormat is the string format this driver uses to store
+// microsecond-precision time in SQLite in text format.
+const TimeFormat = "2006-01-02 15:04:05.000-0700"
+
+func init() {
+	sql.Register("sqlite3", drv)
+}
+
+// ErrMsg sets the driver to collect the sqlite3_errmsg string on error.
+// This is useful but generates extra string allocations.
+// Turn it off if you have written a program with an SQL error on a hot path.
+var ErrMsg = true
+
+var _ driver.Driver = Driver{}
+var _ driver.DriverContext = Driver{}
+var _ driver.ConnBeginTx = (*conn)(nil)
+var _ driver.StmtExecContext = (*stmt)(nil)
+var _ driver.StmtQueryContext = (*stmt)(nil)
+
+var drv Driver
+
+type Driver struct{}
+
+func (d Driver) Open(name string) (driver.Conn, error) { panic("unused") }
+func (d Driver) OpenConnector(name string) (driver.Connector, error) {
+	return &connector{name: name}, nil
+}
+
+type connector struct {
+	name string
+}
+
+func (c *connector) Driver() driver.Driver { return drv }
+func (c *connector) Connect(ctx context.Context) (driver.Conn, error) {
+	log.Printf("connector.Connect: %s", c.name)
+	flags := sqliteh.SQLITE_OPEN_URI |
+		sqliteh.SQLITE_OPEN_READWRITE |
+		sqliteh.SQLITE_OPEN_CREATE
+	db, err := cgosqlite.Open(c.name, flags, "")
+	if err != nil {
+		e := &Error{
+			Code: sqliteh.Code(err.(sqliteh.ErrCode)),
+			Loc:  "Open",
+		}
+		if db != nil {
+			e.Msg = db.ErrMsg()
+			db.Close()
+		}
+		log.Printf("connector.Connect: err=%v", err)
+		return nil, e
+	}
+	return &conn{db: db}, nil
+}
+
+type conn struct {
+	db    *cgosqlite.DB
+	stmts map[string]*stmt // persisted statements
+}
+
+func (c *conn) Prepare(query string) (driver.Stmt, error) { panic("unused") }
```

Comment:
Should this delegate to prepare() also?

I'm going to assume all the panics in here will get implemented later.

Reply:
The `panic("unused")` are as above, methods that are no longer used but we need to meet the types. In this case, a `*conn` is a `driver.Conn` because that's what a `driver.Connector` returns, even though the sql package doesn't use the old methods, only the newer ones from `driver.ConnPrepareContext`.

I found navigating the database/sql/driver package types to be a very un-Go-like experience. :-/